wiring/inc/spark_wiring_ble.h

```diff
@@ -0,0 +1,460 @@
+        0x2d, 0x49, 0xc0, 0x26, 0xf0, 0xdb, 0xce, 0x06, 0x7a, 0x33, 0x6f, 0x12, 0x00, 0x00, 0xf3, 0x5d
+    };
+}
+
+typedef uint16_t BleConnectionHandle;
+typedef uint16_t BleAttributeHandle;
+
+typedef void (*OnDataReceivedCallback)(const uint8_t* data, size_t len, const BlePeerDevice& peer, void* context);
+typedef void (*OnScanResultCallback)(const BleScanResult* device);
+typedef void (*OnConnectedCallback)(const BlePeerDevice& peer);
+typedef void (*OnDisconnectedCallback)(const BlePeerDevice& peer);
+
+class BleAddress : public hal_ble_addr_t {
+public:
+    void operator = (hal_ble_addr_t addr) {
+        this->addr_type = addr.addr_type;
+        memcpy(this->addr, addr.addr, BLE_SIG_ADDR_LEN);
+    }
+
+    uint8_t operator [] (uint8_t i) const {
+        if (i >= BLE_SIG_ADDR_LEN) {
+            return 0;
+        }
+        return addr[i];
+    }
+
+    bool operator == (const BleAddress& addr) const {
+        if (this->addr_type == addr.addr_type && !memcmp(this->addr, addr.addr, BLE_SIG_ADDR_LEN)) {
+            return true;
+        }
+        return false;
+    }
+};
+
+class BleAdvertisingParams : public hal_ble_adv_params_t {
+};
+
+class BleConnectionParams : public hal_ble_conn_params_t {
+};
+
+class BleScanParams : public hal_ble_scan_params_t {
+};
+
+class BleCharacteristicHandles : public hal_ble_char_handles_t {
+public:
+    const BleCharacteristicHandles& operator = (const hal_ble_char_handles_t& halHandles) {
+        this->decl_handle = halHandles.decl_handle;
+        this->value_handle = halHandles.value_handle;
+        this->user_desc_handle = halHandles.user_desc_handle;
+        this->cccd_handle = halHandles.cccd_handle;
+        this->sccd_handle = halHandles.sccd_handle;
+        return *this;
+    }
+};
+
+
+class BleUuid {
+public:
+    BleUuid();
+    BleUuid(const BleUuid& uuid);
+    BleUuid(const uint8_t* uuid128, BleUuidOrder order=BleUuidOrder::LSB);
+    BleUuid(uint16_t uuid16, BleUuidOrder order=BleUuidOrder::LSB);
+    BleUuid(const uint8_t* uuid128, uint16_t uuid16, BleUuidOrder order=BleUuidOrder::LSB);
+    BleUuid(const String& uuid);
+    BleUuid(const char* uuid);
+    ~BleUuid();
+
+    bool isValid(void) const;
+
+    BleUuidType type(void) const {
+        return type_;
+    }
+
+    BleUuidOrder order(void) const {
+        return order_;
+    }
+
+    uint16_t shorted(void) const {
+        return shortUUID_;
+    }
+
+    void full(uint8_t uuid128[BLE_SIG_UUID_128BIT_LEN]) const {
+        memcpy(uuid128, fullUUID_, BLE_SIG_UUID_128BIT_LEN);
+    }
+
+    const uint8_t* full(void) const {
+        return fullUUID_;
+    }
+
+    bool operator == (const BleUuid& uuid) const;
+
+private:
+    BleUuidType type_;
+    BleUuidOrder order_;
+    uint16_t shortUUID_;
+    uint8_t fullUUID_[BLE_SIG_UUID_128BIT_LEN];
+
+    int8_t toInt(char c);
+    void setUuid(const String& str);
+};
+
+
+class iBeacon {
+public:
+    uint16_t major;
```

Comment:
Should these be public?

Reply:
Being public so that user can directly set these members without calling setters separately. It's more like a data structure actually.